src/Java.Interop/Java.Interop/JavaPeerableExtensions.cs

```diff
@@ -11,5 +12,40 @@ public static class JavaPeerableExtensions {
 			JniPeerMembers.AssertSelf (self);
 			return JniEnvironment.Types.GetJniTypeNameFromInstance (self.PeerReference);
 		}
+
+		public static bool TryJavaCast<
+				[DynamicallyAccessedMembers (JavaObject.ConstructorsAndInterfaces)]
+				TResult
+		> (this IJavaPeerable? self, [NotNullWhen (true)] out TResult? result)
+			where TResult : class, IJavaPeerable
+		{
+			result = JavaAs<TResult> (self);
+			return result != null;
+		}
+
+		public static TResult? JavaAs<
+				[DynamicallyAccessedMembers (JavaObject.ConstructorsAndInterfaces)]
+				TResult
+		> (this IJavaPeerable? self)
+			where TResult : class, IJavaPeerable
+		{
+			if (self == null) {
+				return null;
+			}
+
+			if (self is TResult result) {
+				return result;
+			}
+
+			if (!self.PeerReference.IsValid) {
+				throw new ObjectDisposedException (self.GetType ().FullName);
```

Comment:
There are still cases where we throw exceptions (here and in `JniRuntime.JniValueManager.cs`).  I assume these are corner cases where the object is in a "bad state", but I think I would lean towards never throwing an exception under any circumstance.  If the object is bad and we return `null`, the user cannot use it anyways.

Reply:
I mentally split this into two "important" bits:

 1. Java type checking, and
 2. Things that could *never* succeed, and would indicate a fundamental error.

The exceptions are for (2).  For example, what if someone wanted to use this to convert a Java type to `System.ICloneable`:

```csharp
var wat = new Java.Lang.Object().JavaAs<System.ICloneable>();
```

This would fail to *compile* right now, due to the generic constraints on `JavaAs()` (yay), but if someone avoided those constraints (e.g. directly hit `JniEnvironment.Runtime.ValueManager.CreatePeer()`, use `MemberInfo.MakeGenericMethod()` at runtime to create an invalid instantiation {untested; maybe that'll throw?}), then I think it would be better to always throw, because the usage is bad.